input/blog/2018-05-31-cake-v0.28.0-released.md

```diff
@@ -0,0 +1,196 @@
+---
+title: Cake v0.28.0 released
+category: Release Notes
+author: devlead
+---
+
+Version 0.28.0 of Cake has been released.
+
+We're really excited by this release which has some splendid features and improvements!
+
+In this blog post we'll highlight:
+
+- Typed context
+- Setup context
+- Criteria messages
+- "Potentially breaking"
+- Issues & workarounds
+
+Contributions were included from:
+
+- [mholo65](https://github.com/mholo65)
+- [gep13](https://github.com/gep13)
+- [patriksvensson](https://github.com/patriksvensson)
+- [daveaglick](https://github.com/daveaglick)
+- [devlead](https://github.com/devlead)
+
+<!--excerpt-->
+
+### Typed context
+
+In previous versions of Cake you've had to rely on global variables or static members to share variables/state between tasks; this leads to unnecessary complexity and hard to maintain code.
+
+In this release we have introduced the concept of typed context, where in the `Setup` method you can create and return a class; this object is then available to be used in `Teardown` method, Task actions and criteria.
+
+The class can be any C# class, as it's just standard C# you can provide `get` & `set` for properties you want to be mutable, and only get for read only properties.
+
+```csharp
+public class MyBuildData
+{
+	public string Configuration { get; }
+	public bool SignAssemblies { get; }
+	public bool AssmbliesSigned { get; set; }
+
+	public MyBuildData(
+		string configuration,
+		bool signAssemblies)
+	{
+		Configuration = configuration;
+		SignAssemblies = signAssemblies;
+	}
+}
+```
+
+Providing your script with a typed context is done by returning an instance of a class from your script `Setup` method.
+
+```csharp
+Setup<MyBuildData>(setupContext =>
+{
+	return new MyBuildData(
+		configuration: Argument("configuration", "Release"),
+		signAssemblies: BuildSystem.IsRunningOnAppVeyor);
+});
+```
+
+Accessing your data in your task action and criteria, is then done by supplying a type parameter to your task `Does` / `DoesForEach` / `WithCriteria` methods.
+
+```csharp
+Task("Sign")
+    .WithCriteria<MyBuildData>((data, context) => data.SignAssemblies)
```

Comment:
@patriksvensson I have just noticed in this sample code that the parameters `data, context` are different in order compared to the `teardownContext, data` in the Teardown method.  Is there a reason that they are round the other way?  Maybe a bit late in the day, but would it be worth turning them around so that they are consistent?

Reply:
Now `Func<ICakeContext, TData, bool>`